src/trackedit/internal/au3/au3interaction.cpp

```diff
@@ -891,6 +893,145 @@ void Au3Interaction::duplicateTracks(const TrackIdList& trackIds)
+
+void Au3Interaction::moveTracksToTop(const TrackIdList& trackIds)
+{
+    for (const auto trackId : trackIds) {
+        moveTrack(trackId, TrackMoveDir::Top);
+    }
+    projectHistory()->pushHistoryState("Move track", "Move track");
+}
+
+void Au3Interaction::moveTracksToBottom(const TrackIdList& trackIds)
+{
+    for (const auto trackId : trackIds) {
+        moveTrack(trackId, TrackMoveDir::Bottom);
+    }
+    projectHistory()->pushHistoryState("Move track", "Move track");
+}
+
+void Au3Interaction::moveTracksTo(const TrackIdList& trackIds, int to)
+{
+    for (const auto trackId : trackIds) {
+        moveTrackTo(trackId, to++);
+    }
+    projectHistory()->pushHistoryState("Move track", "Move track");
+}
+
+void Au3Interaction::moveTrackTo(const TrackId trackId, int to)
+{
+    auto& project = projectRef();
+    auto& tracks = ::TrackList::Get(project);
+    ::Track* au3Track = DomAccessor::findTrack(project, ::TrackId(trackId));
+
+    IF_ASSERT_FAILED(au3Track) {
+        return;
+    }
+
+    int from = std::distance(tracks.begin(), tracks.Find(au3Track));
+
+    if (to == from) {
+        return;
+    }
+
+    int pos = from;
+    if (pos < to) {
+        while (pos != to && tracks.CanMoveDown(*au3Track)) {
+            pos++;
+            tracks.MoveDown(*au3Track);
+        }
+    } else {
+        while (pos != to && tracks.CanMoveUp(*au3Track)) {
+            pos--;
+            tracks.MoveUp(*au3Track);
+        }
+    }
+
+    if (pos != to) {
+        LOGW() << "Can't move track to " << to;
+    }
+
+    auto track = DomConverter::track(au3Track);
+
+    trackedit::ITrackeditProjectPtr trackEdit = globalContext()->currentTrackeditProject();
+    trackEdit->notifyAboutTrackMoved(track, pos);
+}
+
+void Au3Interaction::moveTrack(const TrackId trackId, TrackMoveDir dir)
+{
+    auto& project = projectRef();
+    auto& tracks = ::TrackList::Get(project);
+    ::Track* au3Track = DomAccessor::findTrack(project, ::TrackId(trackId));
+
+    IF_ASSERT_FAILED(au3Track) {
+        return;
+    }
+
+    int from = std::distance(tracks.begin(), tracks.Find(au3Track));
+
+    IF_ASSERT_FAILED(from >= 0 && from < static_cast<int>(tracks.Size())) {
+        return;
+    }
+
+    int to = from;
+
+    switch (dir) {
+    case TrackMoveDir::Up:
+        if (!tracks.CanMoveUp(*au3Track)) {
+            break;
+        }
+        to--;
+        tracks.MoveUp(*au3Track);
+        break;
+    case TrackMoveDir::Down:
+        if (!tracks.CanMoveDown(*au3Track)) {
+            break;
+        }
+        to++;
+        tracks.MoveDown(*au3Track);
+        break;
+    case TrackMoveDir::Top:
+        while (tracks.CanMoveUp(*au3Track)) {
+            to--;
+            tracks.MoveUp(*au3Track);
+        }
+        break;
+    case TrackMoveDir::Bottom:
+        while (tracks.CanMoveDown(*au3Track)) {
+            to++;
+            tracks.MoveDown(*au3Track);
+        }
+        break;
+    default:
+        return;
+    }
+
+    if (from == to) {
```

Comment:
I don't understand this condition, and also &dir
Could you please explain?

Reply:
Perhaps it is a bad naming on my side. `from` is our starting position, and `to` is resulting position. If we finished at the same position where we started that would mean we failed to move anywhere :)

Will try to come up with better name for those variables

Reply:
@Eism is it more understandable now? :)

Reply:
yes, thanks